### PR TITLE
Add baredSC

### PR DESCRIPTION
**Pull request by lldelisle**
FOR CONTRIBUTOR:
* [X] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [X] - License permits unrestricted use (educational + commercial)
* [X] - This PR adds a new tool or tool collection

### CODE REVIEW

**Comment by lldelisle** [comment]
Ready for review.

File: tools/baredsc/macros.xml [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+            <when value="anndata">
+                <param argument="--inputAnnData" type="data" format="anndata" label="AnnData containing raw counts" />
+            </when>
+        </conditional>
+    </xml>
+    <xml name="macro_single_gene">
+        <param argument="--geneColName" type="text" value="" label="Name of the column with gene counts."/>
+    </xml>
+    <xml name="macro_two_genes">
+        <param argument="--geneXColName" type="text" value="" label="Name of the column with gene counts for gene in x."/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:60]: Where do you get this name from?

File: tools/baredsc/macros.xml [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        <param argument="--outputs" type="data" format="npz" label="Numpy archives from baredSC_@D@d with different number of Gaussians." multiple="true"/>        
+    </xml>
+    <xml name="macro_plots">
+        <param name="image_file_format" type="select" label="Image output format">
+            <option value="png">png</option>
+            <option value="svg">svg</option>
+            <option value="pdf">pdf</option>
+        </param>
+        <param argument="--title" type="text" value="" label="Title to set to all figures."/>
+        <param argument="--removeFirstSamples" type="integer" value="-1" label="Number of samples to ignore before making the plots." help="Use -1 to use a fourth of the number of samples"/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:135]:
```suggestion
        <param argument="--removeFirstSamples" type="integer" value="-1" label="Number of samples to ignore before making the plots" help="Use -1 to use a fourth of the number of samples"/>
```

File: tools/baredsc/macros.xml [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        <param argument="--nsampInPlot" type="integer" value="100000" label="Approximate number of samples to use in plots"/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:136]:
```suggestion
        <param argument="--nsampInPlot" type="integer" value="100000" min="1" label="Approximate number of samples to use in plots"/>   
```

File: tools/baredsc/macros.xml [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+            <option value="png">png</option>
+            <option value="svg">svg</option>
+            <option value="pdf">pdf</option>
+        </param>
+        <param argument="--title" type="text" value="" label="Title to set to all figures."/>
+        <param argument="--removeFirstSamples" type="integer" value="-1" label="Number of samples to ignore before making the plots." help="Use -1 to use a fourth of the number of samples"/>
+        <param argument="--nsampInPlot" type="integer" value="100000" label="Approximate number of samples to use in plots"/>   
+    </xml>
+    <xml name="macro_prettybins_1d">
+        <param argument="--prettyBins" type="integer" value="-1" label="Number of bins to use in plots." help="Use -1 to use the number of bins used in MCMC"/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:139]:
```suggestion
        <param argument="--prettyBins" type="integer" value="-1" min="-1" label="Number of bins to use in plots." help="Use -1 to use the number of bins used in MCMC"/>   
```

File: tools/baredsc/macros.xml [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        </param>
+        <param argument="--title" type="text" value="" label="Title to set to all figures."/>
+        <param argument="--removeFirstSamples" type="integer" value="-1" label="Number of samples to ignore before making the plots." help="Use -1 to use a fourth of the number of samples"/>
+        <param argument="--nsampInPlot" type="integer" value="100000" label="Approximate number of samples to use in plots"/>   
+    </xml>
+    <xml name="macro_prettybins_1d">
+        <param argument="--prettyBins" type="integer" value="-1" label="Number of bins to use in plots." help="Use -1 to use the number of bins used in MCMC"/>   
+    </xml>
+    <xml name="macro_prettybins_axis" token_axis="x">
+        <param argument="--prettyBins@AXIS@" type="integer" value="-1" label="Number of bins to use in @AXIS@ in plots." help="Use -1 to use the number of bins used in MCMC"/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:142]:
```suggestion
        <param argument="--prettyBins@AXIS@" type="integer" value="-1" min="-1" label="Number of bins to use in @AXIS@ in plots." help="Use -1 to use the number of bins used in MCMC"/>   
```

File: tools/baredsc/macros.xml [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        <param argument="--nsampInPlot" type="integer" value="100000" label="Approximate number of samples to use in plots"/>   
+    </xml>
+    <xml name="macro_prettybins_1d">
+        <param argument="--prettyBins" type="integer" value="-1" label="Number of bins to use in plots." help="Use -1 to use the number of bins used in MCMC"/>   
+    </xml>
+    <xml name="macro_prettybins_axis" token_axis="x">
+        <param argument="--prettyBins@AXIS@" type="integer" value="-1" label="Number of bins to use in @AXIS@ in plots." help="Use -1 to use the number of bins used in MCMC"/>   
+    </xml>
+    <xml name="macro_splity">
+        <param argument="--splity" type="text" value="" label="Threshold values separated by space to plot the density for genex for 2 categories in geney values." help="Leave empty if you don't need this type of analysis.">
```
Comment (bgruening) [inline tools/baredsc/macros.xml:145]:
```suggestion
        <param argument="--splity" type="text" value="" label="Threshold values separated by space to plot the density for genex for 2 categories in geney values" help="Leave empty if you don't need this type of analysis.">
```

File: tools/baredsc/macros.xml [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+            <validator type="regex">(-?[0-9]+( -?[0-9]+)*)?</validator>
+        </param>
+    </xml>
+    <xml name="macro_colorscale">
+        <param argument="--log1pColorScale" type="boolean" truevalue="--log1pColorScale" falsevalue="" checked="false" help="This helps to see regions in plot with low proportion of cells."/>
```
Comment (bgruening) [inline tools/baredsc/macros.xml:150]:
```suggestion
        <param argument="--log1pColorScale" type="boolean" truevalue="--log1pColorScale" falsevalue="" checked="false" label="Enable to see regions in plot with low proportion of cells"/>
```

File: tools/baredsc/baredsc_1d.xml [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+            <expand macro="macro_advanced_common_baredSC"/>
+        </section>
+    </inputs>
+    <outputs>
+        <data format="npz" name="output" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneColName}: Numpy archive" from_work_dir="output.npz"/>
+        <data format="txt" name="neff" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneColName}: Neff" from_work_dir="output/baredSC_neff.txt"/>
+        <collection type="list" name="qc_plots" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneColName}: QC plots">
+            <discover_datasets pattern="baredSC_(?P&lt;identifier_0&gt;\S+)\.(?P&lt;ext&gt;.*)" directory="QC"/>
+        </collection>
+        <data format="txt" name="pdf" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneColName}: pdf with error bar" from_work_dir="output/baredSC_pdf.txt"/>
```
Comment (bgruening) [inline tools/baredsc/baredsc_1d.xml:67]: this looks more like a tabular isn't it?

File: tools/baredsc/baredsc_1d.xml [hunk 9]
```diff
@@ -0,0 +1,298 @@
...
+        <data format="txt" name="logevidence" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneColName}: Log evidence" from_work_dir="logevidence.txt"/>
+    </outputs>
+    <tests>
+        <!-- First test -->
+        <test>
+            <conditional name="input_counts">
+                <param name="filetype" value="tabular"/>
+                <param name="input" value="nih3t3_generated_2d_2.txt"/>
+            </conditional>
+            <param name="geneColName" value="0.5_0_0_0.5_x"/>
```
Comment (bgruening) [inline tools/baredsc/baredsc_1d.xml:87]: where do you get those column names from?

File: tools/baredsc/baredsc_1d.xml [hunk 10]
```diff
@@ -0,0 +1,298 @@
...
+                        <has_size value="45302" delta="4000"/>
+                    </assert_contents>
+                </element>
+                <element name="corner" ftype="png">
+                    <assert_contents>
+                        <has_size value="66254" delta="6000"/>
+                    </assert_contents>
+                </element>
+            </output_collection>
+            <output name="pdf">
```
Comment (bgruening) [inline tools/baredsc/baredsc_1d.xml:121]: ftype="txt" or tabular .... maybe a comment that pdf means probability density function.

File: tools/baredsc/baredsc_1d.xml (hunk 8 above)
Comment (bgruening) [reply tools/baredsc/baredsc_1d.xml:67]: You could add the column names as metadata here x\s+low\s+mean\s+high\s+median

File: tools/baredsc/baredsc_2d.xml [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+            <expand macro="macro_scaleprior"/>
+        </section>
+    </inputs>
+    <outputs>
+        <data format="npz" name="output" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneXColName} and ${geneYColName}: Numpy archive" from_work_dir="output.npz"/>
+        <data format="txt" name="neff" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneXColName} and ${geneYColName}: Neff" from_work_dir="output/baredSC_neff.txt"/>
+        <collection type="list" name="qc_plots" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneXColName} and ${geneYColName}: QC plots">
+            <discover_datasets pattern="baredSC_(?P&lt;identifier_0&gt;\S+)\.(?P&lt;ext&gt;.*)" directory="QC"/>
+        </collection>
+        <data format="txt" name="pdf2d" label="${tool.name} on $getVar('input_counts.input', 'input_counts.inputAnnData') for ${geneXColName} and ${geneYColName}: mean pdf in matrix format" from_work_dir="output/baredSC_pdf2d.txt"/>
```
Comment (bgruening) [inline tools/baredsc/baredsc_2d.xml:75]: tabular?

File: tools/baredsc/baredsc_combine_1d.xml [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+        <edam_operation>operation_2495</edam_operation>
+    </edam_operations>
+    <xrefs>
+        <xref type="bio.tools">baredsc</xref>
+    </xrefs>
+    <expand macro="requirements"/>
+    <command detect_errors="exit_code"><![CDATA[
+    ## Create symlinks
+    #for $i, $output in enumerate($MCMC.outputs):
+        ln -s "$output" '$i'.npz &&
```
Comment (bgruening) [inline tools/baredsc/baredsc_combine_1d.xml:17]:
```suggestion
        ln -s '$output' '$i.npz' &&
```

File: tools/baredsc/baredsc_combine_2d.xml [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+        <edam_operation>operation_3463</edam_operation>
+    </edam_operations>
+    <xrefs>
+        <xref type="bio.tools">baredsc</xref>
+    </xrefs>
+    <expand macro="requirements"/>
+    <command detect_errors="exit_code"><![CDATA[
+    ## Create symlinks
+    #for $i, $output in enumerate($MCMC.outputs):
+        ln -s "$output" '$i'.npz &&
```
Comment (bgruening) [inline tools/baredsc/baredsc_combine_2d.xml:18]:
```suggestion
        ln -s '$output' '$i.npz' &&
```

File: tools/baredsc/macros.xml [hunk 14]
```diff
@@ -0,0 +1,313 @@
...
+            
+            To get a description of outputs, please read the `Documentation <https://baredsc.readthedocs.io/en/latest/index.html>`_
+
+            This is a description of the figure with the results. 
+
+            - When the 1d version is used, it displays the mean pdf in solid red line, the median in black dashed lines (/!\backslash the integral of the median is not equal to 1) with the confidence interval of 1 sigma (68%), 2 sigma (95%) and 3 sigma (99.7%) as well as in green, the kernel density estimate of the input values, the detected expression (``log(1 + targetSum * raw / total UMI)``).
```
Comment (bgruening) [inline tools/baredsc/macros.xml:24]: PDF in capital letter or lower letters?

File: tools/baredsc/macros.xml (hunk 1 above)
Comment (lldelisle) [reply tools/baredsc/macros.xml:60]: These must be in the first row (header).

File: tools/baredsc/baredsc_1d.xml (hunk 8 above)
Comment (lldelisle) [reply tools/baredsc/baredsc_1d.xml:67]: Right

File: tools/baredsc/macros.xml (hunk 14 above)
Comment (lldelisle) [reply tools/baredsc/macros.xml:24]: I should use capital letters everywhere.

**Comment by lldelisle** [comment]
Thanks for the review I will incorporate changes later this evening.

**Comment by lldelisle** [comment]
As I could not find a way to set metadata (column_names) for uploaded file, is it ok to keep it like this?

**Comment by bgruening** [comment]
Sure.

**Comment by lldelisle** [comment]
(I don't know why the check workflow success step is systematically cancelled)

**Review by bgruening — APPROVED** [review]

**Pull request merged**